config/hotreloadable.go

```diff
@@ -143,115 +384,304 @@ func (c *Config) RegisterDurationConfigVariable(defaultValueInTimescaleUnits int
+func (c *Config) RegisterStringMapVar(defaultValue map[string]interface{}, orderedKeys ...string) *map[string]interface{} {
+	ptr := getOrCreatePointer(
+		c.vars, c.varsMisuses, &c.varsLock, defaultValue, false, orderedKeys...,
+	).(*map[string]interface{})
+	c.registerStringMapVar(defaultValue, ptr, false, func(v map[string]interface{}) {
+		*ptr = v
+	}, orderedKeys...)
+	return ptr
+}
+
+// RegisterReloadableStringMapVar registers a hot-reloadable string map config variable
+// WARNING: a different order of keys will result in a different variable
+func (c *Config) RegisterReloadableStringMapVar(
+	defaultValue map[string]interface{}, orderedKeys ...string,
+) *Reloadable[map[string]interface{}] {
+	ptr := getOrCreatePointer(
+		c.vars, c.varsMisuses, &c.varsLock, defaultValue, true, orderedKeys...,
+	).(*Reloadable[map[string]interface{}])
+	c.registerStringMapVar(defaultValue, ptr, true, func(v map[string]interface{}) {
+		ptr.store(v)
+	}, orderedKeys...)
+	return ptr
+}
+
+func (c *Config) registerStringMapVar(
+	defaultValue map[string]interface{}, ptr any, isHotReloadable bool, store func(map[string]interface{}), orderedKeys ...string,
+) {
 	c.vLock.RLock()
 	defer c.vLock.RUnlock()
 	c.hotReloadableConfigLock.Lock()
 	defer c.hotReloadableConfigLock.Unlock()
 	configVar := configValue{
 		value:        ptr,
 		defaultValue: defaultValue,
-		keys:         keys,
+		keys:         orderedKeys,
 	}
 
 	if isHotReloadable {
-		c.appendVarToConfigMaps(keys[0], &configVar)
+		c.appendVarToConfigMaps(orderedKeys, &configVar)
 	}
 
-	for _, key := range keys {
+	for _, key := range orderedKeys {
 		if c.IsSet(key) {
-			*ptr = c.GetStringMap(key, defaultValue)
+			store(c.GetStringMap(key, defaultValue))
 			return
 		}
 	}
-	*ptr = defaultValue
+	store(defaultValue)
 }
 
-func (c *Config) appendVarToConfigMaps(key string, configVar *configValue) {
+func (c *Config) appendVarToConfigMaps(keys []string, configVar *configValue) {
+	key := strings.Join(keys, ",")
 	if _, ok := c.hotReloadableConfig[key]; !ok {
 		c.hotReloadableConfig[key] = make([]*configValue, 0)
 	}
 	c.hotReloadableConfig[key] = append(c.hotReloadableConfig[key], configVar)
 }
+
+type configTypes interface {
+	int | int64 | string | time.Duration | bool | float64 | []string | map[string]interface{}
+}
+
+// Reloadable is used as a wrapper for hot-reloadable config variables
+type Reloadable[T configTypes] struct {
+	value T
+	lock  sync.RWMutex
+}
+
+// Load should be used to read the underlying value without worrying about data races
+func (a *Reloadable[T]) Load() T {
+	a.lock.RLock()
+	v := a.value
+	a.lock.RUnlock()
+	return v
+}
+
+func (a *Reloadable[T]) store(v T) {
+	a.lock.Lock()
+	a.value = v
+	a.lock.Unlock()
+}
+
+// swapIfNotEqual is used internally to swap the value of a hot-reloadable config variable
+func (a *Reloadable[T]) swapIfNotEqual(new T, compare func(old, new T) bool) (old T, swapped bool) {
```

Comment:
```suggestion
// swap is used internally to swap the value of a hot-reloadable config variable if the new value is not equal to the old value
func (a *Reloadable[T]) swap(new T, compare func(old, new T) bool) (old T, swapped bool) {
```

Reply:
I'd rather have a more descriptive method name to be honest, but I'll update the comment :+1: